### PR TITLE
Use platform-specific path resolution utilities

### PR DESCRIPTION
**Pull request by alastairs**
A quick spike into resolving #397. The tests fail in NCrunch but pass in
ReSharper's test runner, which is almost certainly to do with NCrunch's
own path isolation.

The approach is to use where.exe on Windows and `which` on *nix systems.
While `which` is well-known, where.exe appears not to be. The help
describes it as a utility which:

> Displays the location of files that match the search pattern.
  By default, the search is done along the current directory and
  in the paths specified by the PATH environment variable.

### CODE REVIEW

**Comment by mwadams** [comment]
I think I would be happy if we had something that was working in the VS test runner and R#, `dotnet test` and R#

I've put it on our PR review queue - and suggest we open a separate issue to get it running under NCrunch.

@idg10 is an NCrunch user, I believe. There may be others!

**Comment by alastairs** [comment]
I'll double-check against those test runners this morning 👍🏻

File: Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs [hunk 1]
```diff
@@ -184,22 +184,21 @@ public void TeardownFunctions()
...
-            string toolsFolder = Path.Combine(
-                npmPrefix,
-                @"node_modules\azure-functions-core-tools\bin");
-
-            if (!Directory.Exists(toolsFolder))
-            {
-                throw new FunctionStartupException(
-                    $"Azure Functions runtime not found at {toolsFolder}. Have you run: " +
-                    "'npm install -g azure-functions-core-tools@3 --unsafe-perm true'?");
-            }
```
Comment (MikeEvansLarah) [inline Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs:197]: We could keep this logic to throw a `FunctionStartupException` if the runtime is not installed.

I think we could check for a non-zero exit code for both `where` and `which` (which would indicate the runtime is not installed)

Comment (alastairs) [reply Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs:197]: Reinstated the exception thrown if the runtime is not installed👍🏻

File: Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs [hunk 2]
```diff
@@ -184,22 +184,21 @@ public void TeardownFunctions()
...
+            string toolLocatorName = Environment.OSVersion.Platform == PlatformID.Win32NT ? "where.exe" : "which";
+            string toolName = Environment.OSVersion.Platform == PlatformID.Win32NT ? "func.exe" : "func";
```
[the remark this answers is not in the document]
Comment (alastairs) [reply Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs]: I've implemented a loop that attempts to invoke each item returned by `where.exe`. I've not been able to test this on non-Windows platforms so far, but based on #271 I _think_ this library is still Windows-only... (this tripped us up on our CI pipelines, but I noted elsewhere you'd seen other difficulties with Functions on non-Windows).

**Comment by alastairs** [comment]
I think I've resolved all comments left in the previous review (thanks @MikeEvansLarah!). This issue is the remaining obstacle blocking us updating our Functions to .NET 6 and Functions v4, so an expedited review next week would be _hugely_ appreciated if you have capacity for it.

**Comment by idg10** [comment]
The tests seem to be failing in the automated build:"

https://dev.azure.com/endjin-labs/Corvus.Testing/_build/results?buildId=20412&view=results

I kicked them to re-run, just in case it was a transient failure—sometimes build agents go weird and we get failures for no good reason. But it looks reproducable.

The first failure is this:

```
[xUnit.net 00:00:15.50]     Corvus.Testing.AzureFunctions.Xunit.Demo.FunctionPerTestFacts.A_Post_request_without_a_query_string_or_request_body_fails [FAIL]
  Failed Corvus.Testing.AzureFunctions.Xunit.Demo.FunctionPerTestFacts.A_Post_request_without_a_query_string_or_request_body_fails [1 ms]
  Error Message:
   System.InvalidOperationException : Cannot start process because a file name has not been provided.
  Stack Trace:
     at System.Diagnostics.Process.Start()
   at Corvus.Testing.AzureFunctions.Internal.ProcessOutputHandler.Start() in /_/Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/Internal/ProcessOutputHandler.cs:line 111
   at Corvus.Testing.AzureFunctions.FunctionsController.GetToolPath() in /_/Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs:line 285
   at Corvus.Testing.AzureFunctions.FunctionsController.StartFunctionHostProcess(Int32 port, String provider, String workingDirectory, FunctionConfiguration functionConfiguration) in /_/Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs:line 316
   at Corvus.Testing.AzureFunctions.FunctionsController.StartFunctionsInstance(String path, Int32 port, String runtime, String provider, FunctionConfiguration configuration) in /_/Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/FunctionsController.cs:line 90
   at Corvus.Testing.AzureFunctions.Xunit.Demo.FunctionPerTestFacts.InitializeAsync() in /_/Solutions/Corvus.Testing.AzureFunctions.Xunit.Demo/FunctionPerTestFacts.cs:line 99
```

This is actually one of the demo projects that we use to show how to use things. These aren't strictly tests of the code themselves, but the are supposed to be working examples, so the fact that they fail is a problem.

We also get a bunch of errors looking like this:

```
{"name":"Corvus.Testing.AzureFunctions.Xunit.Demo (OS: Microsoft Windows 10.0.20348, Job: Test, Agent: Hosted Agent)","build":{"id":"20412"},"startedDate":"2023-09-04T09:44:43.966Z","isAutomated":true}
Exception:
System.Net.Http.HttpRequestException: Response status code does not indicate success: 500 (Internal Server Error).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at AzurePipelines.TestLogger.ApiClient.SendAsync(HttpMethod method, String endpoint, String body, CancellationToken cancellationToken, String apiVersionString)
```

Those look more like something went wrong in the world of Azure DevOps. However, it's the same project failing, so it may be related.

Did you look at the Demo test projects when making these changes? Do they work for you locally?

**Comment by MikeEvansLarah** [comment]
> Those look more like something went wrong in the world of Azure DevOps. However, it's the same project failing, so it may be related.

@JamesDawson and I had a quick look into these errors a couple weeks ago and think they are permissions issues with the Azure DevOps logger when a build is triggered from a fork. We're going to take another look this afternoon.

I don't think these errors are related to the `System.InvalidOperationException : Cannot start process because a file name has not been provided.` ones though

**Comment by alastairs** [comment]
Thanks @idg10 and @MikeEvansLarah for the replies. The `InvalidOperationException` is what is thrown when the `func`-resolution code doesn't return any results, so it's a surprise that it's failing for only some tests and not all. The Demo tests did all pass on my machine, yes, but the culprit looks to be 6e7efbe for reasons I don't yet understand. Is it possible to rerun the pipeline against the previous commit (1f859ad), which appears to have run happily? It would be useful to understand whether the failure is reproducible there too.

**Comment by idg10** [comment]
Sorry for the delay. For some reason the branch in our repo for this PR doesn't appear to picking up changes from your repo, and Azure DevOps won't let us tell it build a branch from your repo, so I had to do some hackery to get to the point where we could actually do what you suggested, and try to build with that last commit reverted.

What I've ended up doing is creating a new branch, https://github.com/corvus-dotnet/Corvus.Testing/tree/test-alastairs-issue-397 which is fully up to date with the branch in your repo, and then doing a revert commit to back out the change you think might be the problem, and then manually kicking off a build for that branch.

That's this build: https://dev.azure.com/endjin-labs/Corvus.Testing/_build/results?buildId=20802&view=results

This didn't help.

I also tried a modification:

```cs
string[] toolPaths = toolLocator.StandardOutputText.Split("\n").Select(s => s.Trim()).Where(s => s.Length > 0).ToArray();
```

I added that `Where` clause to skip any empty lines. I did that because the error message suggested that we were trying to pass an empty string to `ProcessStartInfo`. My guess was that we were for some reason getting no output at all, so I was expecting this change to just run into the exception that gets thrown when none of the tool paths worked.

But to my surprise, the build seemed to pass:

https://dev.azure.com/endjin-labs/Corvus.Testing/_build/results?buildId=20803&view=results

I don't understand why though. I don't know why only that one project would behave differently. I don't know why we'd sometimes get blank lines out of `where.exe` or `which`. So it's a bit unsatisfactory.

Anyway, I made this change on a random branch rigged up to work around the fact that the github-created branch for this PR seems to be behind. But I think if a) you're happy with that last change I described, and b) you make it on your branch, that should trigger another build of this PR, which, unless that last time was a fluke, should pass.

**Comment by alastairs** [comment]
I wonder if there's a race condition or similar happening here which makes this flaky 🤔  I'm suspicious (and expecting a fluke) because:
1. 1f859ad passed and 6e7efbe did not, even though there were no functional changes made
2. A branch equivalent to 1f859ad did _not_ pass
3. We don't understand why the filter allowed that build to pass. 

Another couple of pipeline runs ought to confirm whether or not this is a stable fix, I think? I'll try one now using the magic slash command; I've no idea if it'll actually work for me or not!

**Comment by idg10** [comment]
I ran the build again and it failed in exactly the same way: https://dev.azure.com/endjin-labs/Corvus.Testing/_build/results?buildId=20908&view=results

So I then did it one more time, and this time it didn't fail: https://dev.azure.com/endjin-labs/Corvus.Testing/_build/results?buildId=20909&view=logs&j=5264e576-3c6f-51f6-f055-fab409685f20&t=9f5eba70-d33b-5377-96f2-d94c82946629

So it is intermittent, but the failure wasn't a one-off either because we've seen exactly the same failure at least a couple of times.

So I don't think we can merge this just yet, because it looks like it will fail at least some of the time.

What I've done now is modify my branch with a copy of your changes, the same branch in which I experimented with filtering out empty output lines from `where.exe`, and I've added some diagnostic logging to try and get a better picture of what it's actually trying to do prior to the failure in the hope of getting some insight into why we sometimes see these errors. You can see output at https://dev.azure.com/endjin-labs/Corvus.Testing/_build/results?buildId=20910&view=logs&j=5264e576-3c6f-51f6-f055-fab409685f20&t=9f5eba70-d33b-5377-96f2-d94c82946629&l=72

In particular:

```
'C:\npm\prefix\func', 'C:\npm\prefix\func.cmd', '', ''
'C:\npm\prefix\func', 'C:\npm\prefix\func.cmd', '', ''
Testing tool path 'C:\npm\prefix\func' can be used for running Functions projects.
Testing tool path 'C:\npm\prefix\func' can be used for running Functions projects.
Testing tool path 'C:\npm\prefix\func.cmd' can be used for running Functions projects.
Testing tool path 'C:\npm\prefix\func.cmd' can be used for running Functions projects.
'C:\npm\prefix\func', 'C:\npm\prefix\func.cmd', '', ''
Testing tool path 'C:\npm\prefix\func' can be used for running Functions projects.
Testing tool path 'C:\npm\prefix\func.cmd' can be used for running Functions projects.
''
[xUnit.net 00:00:19.90]     Corvus.Testing.AzureFunctions.Xunit.Demo.FunctionPerTestFacts.A_Post_request_without_a_query_string_or_request_body_fails [FAIL]
Testing tool path '' can be used for running Functions projects.
'C:\npm\prefix\func', 'C:\npm\prefix\func.cmd', '', ''
Testing tool path 'C:\npm\prefix\func' can be used for running Functions projects.
Testing tool path 'C:\npm\prefix\func.cmd' can be used for running Functions projects.
```

The first two lines are both showing output produced with this code:

```cs
string.Join(", ", toolPaths.Select(p => $"'{p}'")
```

We see it multiple times because there are multiple test projects. The first two lines are for two of the three test projects, and the third appears a bit further down:

```
''
```

So there are two things to note:

1. `where.exe` includes blank lines in its output, so your code `toolLocator.StandardOutputText.Split("\n").Select(s => s.Trim()).ToArray()` is indeed a problem. (I removed my temporary fix that filtered out the blank lines here to test whether that was indeed happening. It is.)
2. Apparently when we run in one of the three test runners, `where.exe` fails to find `func`

I'm pretty sure 2 is happening in the xunit test project, but I've added some additional debug output to try to verify this. (Two of the three projects use unit, and two of the three projects find `func` without problems, so it seems likely there's something different about the xunit runner.)

So I'm now wondering if the xunit runner is changing the `PATH` for some reason. I've also added debug spew to try to capture the path. (These builds take a while so I don't have the results yet.)

**Comment by idg10** [comment]
Ah... I have a new hypothesis. I think the problem might actually be this line here:

https://github.com/corvus-dotnet/Corvus.Testing/blob/main/Solutions/Corvus.Testing.AzureFunctions/Corvus/Testing/AzureFunctions/Internal/ProcessOutputHandler.cs#L197

I think it's possible that we get the `Process.Exited` event before we've finished receiving all output, and so sometimes we detach the output event handler before we've received all the output from `where.exe` which could explain why we sometimes end up with what looks like empty output.

It initially seemed odd to me that we seem to get a single blank line rather than no output at all, but apparently calling `Split("\n")` on an empty string produces an array containing a single empty string.


OK... so here's what I think needs to be changed:

1. I think you should add the `Where(s => s.Length > 0)` clause I suggested earlier to remove empty entries from `toolPaths`, because we are _definitely_ seeing empty entries in that list
2. remove that line of code from the `ProcessOutputHandler` to avoid a race condition in which we sometimes don't receive all of the output from `where.exe`, which I think is what is causing these sporadic failures

**Comment by idg10** [comment]
OK, since the `ProcessOutputHandler` issue was ours, we've fixed that in #410 

That's now merged into this branch and all checks pass. I'm going to create a separate issue to deal with my suggestion of adding that `Where(s => s.Length > 0)` suggested above.

**Review by idg10 — APPROVED** [review]

**Pull request merged**